src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.SymbolKeyReader.cs

```diff
@@ -493,9 +576,24 @@ private SymbolKeyResolution ReadWorker(SymbolKeyType type, out string? failureRe
             /// Callers should <see cref="IDisposable.Dispose"/> the instance returned.  No check is
             /// necessary if <c>default</c> was returned before calling <see cref="IDisposable.Dispose"/>
             /// </summary>
-            public PooledArrayBuilder<TSymbol> ReadSymbolKeyArray<TSymbol>(out string? failureReason) where TSymbol : ISymbol
+            public PooledArrayBuilder<TSymbol> ReadSymbolKeyArray<TContextualSymbol, TSymbol>(
+                TContextualSymbol? contextualSymbol,
+                Func<TContextualSymbol, int, ITypeSymbol?> getContextualType,
+                out string? failureReason)
+                where TContextualSymbol : ISymbol
+                where TSymbol : ISymbol
             {
-                using var resolutions = ReadArray(_readSymbolKey, out var elementsFailureReason);
+                using var resolutions = ReadArray(
+                    _readSymbolKey,
+                    static (data, i) =>
+                    {
+                        var contextualType = data.contextualSymbol is null ? null : data.getContextualType(data.contextualSymbol, i);
+                        data.self.PushContextualType(contextualType);
+                        return (data.self, contextualType);
+                    },
+                    static (data) => data.self.PopContextualType(data.contextualType),
+                    (self: this, contextualSymbol, getContextualType),
+                    out var elementsFailureReason);
```

Comment:
it's ok if your brain has given up on this stuff.  primary goal is zero allocs, while also having a uniform way of processing arrrays of items in the SymbolKeys.

Reply:
ok, moved this into a separate method.  this is too complex.